### DISCUSSION TITLE
Meta workflow manager

### DISCUSSION THREAD

**Discussion by lecorguille** [IFB-ElixirFr/EMERGEN-workflows#11] [Ideas]
Report de la discussion du 26 mai : 

----

@lecorguille  10 h 48

> Tant que je vous tiens. Dans le cadre d’ermergen, nous aurons des jobs ci a lancer régulièrement (et pas depuis des repo gitlab)
> Quand pensez-vous d’instancier un serveur jenkins ? On pourrait aussi du coup y faire tourner les tests de workflow. C’est en effet peut-être plus pratique pour le suivi que les jobs périodiques de gitlab-ci ?

@FanchTheSystem   10 h 49

> Jenkins apporterai quoi ?
> A mon avis Jenkins c'est un grand pas en arriere comparer au fonctionnalité des runner gitlab.
> On va galerer a le maintenir, c'est pas vraiment top d'un point de vue sécurité (à moins d'etre completement isolé d'internet) et c'est trés figé dans la facon de faire (genre si on fait pas comme c'est prevu, c'est presque impossible à gerer/adapter)
> C'est vraiment un outil qui a mal veilli
> 

@lecorguille  12 h 22

> Je me dis que les gitlab-ci permet de lancer des jobs depuis gitlab-ci.
> Pour emergen, c’est la détection de nouveaux fichiers sur le FS qui déclenchera des jobs (sous Galaxy). Est-ce les gitlab-ci sont fait pour ça ?
> De plus, il serait intéressant de proposer une interface de monitoring pour les partenaires et les gitlab-ci c’est pas super ergonomique, je trouve.

Guillaume  13 h 40

> ça m'intéresse mais je ne suis pas suffisamment calé sur les outils d'orchestration. Si c'est un juste une détection de nouveaux fichiers pour lancer des jobs, est-ce que ce n'est pas plus intéressant d'écrire un outil qui fait ça directement dans emergen (check fs, run job) plutôt que de rajouter une grosse brique supplémentaire?

@FanchTheSystem   13 h 41

> Gitlab runner permet de lancer des jobs depuis d'autre gestionnaire de source (githyb bitbucket ...) et Jenkins c'est pareil, dans les deux cas c'est des outils fait pour declencher des jobs lors de changement dans un code source.
> Aucun des deux n'est adapter pour detecter des nouveaux fichier, et c'est possible de le faire seulement en detournant leur usage.
> Personnelement je trouve pas l'interface de jenkins ergonomique, je la trouve meme un peu horrible (10 ans de retard).
> Peut-etre qu'en fait ca vaux le coup de chercher un peu ce qui pourrait etre adapter au besoin plutot que de prendre le premier truc qui vient et essayer de le bricoler.
> En gros pour faire "bien" il faudrait bien reflechir au besoin, definir quel brique fonctionnel on veux, et faire un peu de veille pour voir ce qui colle. Soit plusieurs petites briques, soit une grosse.
> Mais dans tout les cas, eviter de partir de la solution proposé pour definir le besoin.
> Et oui ca prend un peu de temps de faire ca :visage_légèrement_souriant:
> Je rajoute une couche et j'arrete aprés :clin_d'œil:
> Jenkins c'est fait pour declencher des builds d'application Java quand une modification est faite sur un code source.
> Faire autre chose que des builds, ou essayer de build d'autre langage, ou declencher sur d'autre type d'evenement ... Tout ca c'est possible mais en bricolant, et en interne d'un point de vue technique c'est souvent trés crade.
> Ca en valait la peine quand c'etait un peu le seul truc correct pour faire ce genre de chose (il y'a 10 ans).
> Mais aujourd'hui avec un peu d'effort j'ai aucun doute sur le fait qu'on peux faire 1000 fois mieux.
> Un exemple d'alternative on pourrait faire un detecteur de changement au niveau du systeme de fichier (avec systemd ou autre) et proposer un dashboard de resultat avec grafana

@julozi   13 h 54

> Il est aussi possible de lancer des jobs GitLab CI hors integration continue
> Sous la forme de job programmé ou sur évènement de l’API notamment

@lecorguille   14 h 01

> Je disais Jenkins parce que c’est en effet l’un des plus connu et parce que c’est celui utiliser par les :uk: dans le même context. Je suis ouvert à toutes propositions.
> Et oui, j’avais aussi envisager de coupler ça avec un Grafana parce que la com’ peut aussi être le nerf de la guerre.
> Galaxy peut aussi servir pour les partenaires avec l’interface d’invocation des workflows.
> Donc si gitlab-ci peut faire l’affaire couplé à une visu plus friendly, ça peut le faire.

@julozi   14 h 02

> En gros on cherche un truc pour lancer automatiquement des workflow sur galaxy quand des données arrivent, c’est ça ?

@lecorguille   14 h 02

> C’est ça

@julozi  14 h 03

> Pas sûr que GitLab-ci est du sens dans ce contexte

@lecorguille  14 h 03

> C’est ce que je me disais

@julozi  14 h 03

> L’idée d’utiliser un événement systemd + suivi via grafana semble tip-top

@lecorguille  14 h 06

> A priori CircleCi est payant : https://circleci.com/product/#hosting-options

@FanchTheSystem  14 h 06

> mais pourquoi tu veux absolument un outil d'integration continue pour faire autre chose que de l'integration continue ?

@lecorguille  14 h 07
> Keep calm !

@lecorguille  14 h 15
> Le workflow :
> 1. Détection de nouveaux fichiers sur le FS
> 2. Workflow dans le workflow
>   a. Nettoyage de données humaines [hors Galaxy] ?
>   b. Workflow Galaxy
> 3. Downstream : email, rapport ?

Il y a plusieurs acteurs :
> - Nous/Admin : on veut une interface ou truc pour voir que la communication entre le FS et Galaxy se fait bien. Nous allons  peut-être aussi avoir des steps en amont et aval de Galaxy (c’est à définir) qu’il faudra “tracer”
> - Les partenaires : peut-être que l’interface Galaxy peut les contenter
> - Le “public” : un grafana peut faire l’affaire
> J’aime bien l’idée d’avoir un dash board avec le suivi, des barres colorés qui se remplissent, du vert, du rouge, des alertes si rouge, la possibilité de relancer si besoin ... (modifié) 

@julozi  14 h 16

> Vu que la partie galaxy n’est qu’une étape du workflow complet, ça serait cool d’avoir un outil chez nous pour définir et piloter l’exécution de ce workflow
> (mais pas un outil de ci hein, je ne suis qu’amour)

@julozi  14 h 20

> Est-ce qu’on serait pas un peu en train de chercher un outil genre ETL en fait ?
> https://github.com/pawl/awesome-etl

@lecorguille  14 h 59

> Il n’y a pas Galaxy :cri:
> De la liste, j’ai trouvé AirFlow ou Azkaban mais sans plus creuser.

@julozi  15 h

> Galaxy est spécialisé pour la science,  mais on pourrait le voir comme un ETL à certain niveau effectivement

@lecorguille  14 h 39

> Profitons que le sujet est chaud
> Un avis pour https://airflow.apache.org/docs/apache-airflow/stable/ui.html ?
> (ping @Christophe Antoniewski, un petit thread en rapport avec Emergen-Galaxy)

**Comment by lecorguille** [comment]
# Airflow

J'ai un peu regarder https://airflow.apache.org/ et cela semble pas mal.

Après avoir regarder cette petite vidéo [Airflow for Beginners - Run Spotify ETL Job in 15 minutes!](https://www.youtube.com/watch?v=i25ttd32-eo). Je l'ai installé rapidement avec ce [Quick Start](https://airflow.apache.org/docs/apache-airflow/stable/start/local.html). 
On peut voir rapidement ce que cela donne.

Je vois pas mal de features qui pourraient nous intéresser (si j'ai bien tout compris):
 - DAG avec des tasks bash ou python
 - Scheduler à la cron
 - Possibilité de créer des utilisateurs et des roles et leur limité les DAG visible
 - Affichage des tasks avec des status success et failed
 - Vue Calendrier
 - Email en cas d'échec, ...
 - Possibilité de monitorer avec statsd et donc grafana
 - ...

Un avis ?


Quelques liens de doc que je trouve utile : 
 - https://www.astronomer.io/guides/airflow-scaling-workers
 - https://www.astronomer.io/guides/dynamically-generating-dags
 - https://stackoverflow.com/questions/60082546/airflow-proper-way-to-run-dag-for-each-file
 - https://medium.com/@igorlubimov/dynamic-scheduling-in-airflow-52979b3e6b13

**Reply by julozi** [reply]
J'ai découvert Airflow au travers de la vidéo que tu as mis en lien.
L'outil semble vraiment prometeur et permettrait d'avoir une regard clair non seulement sur le contenu des workflows mais aussi sur les executions réalisées ou en cours.
La définition du DAG au travers de Python semble également idéal pour un interfaçage avec Bioblend.
Je pense que l'on pourrait très facilement monter une maquette sur une VM et voir où ça nous mène.

**Comment by julozi** [comment]
J'ai fait un peu de veille pour trouver une solution propre pour déclencher un traitement (script) sur l'arrivée de nouvelles données sur le stockage.
L'approche la plus simple serait d'utiliser la solution `incron` qui permet de déclencher l'éxecution d'un script sur un évènement inotify pour un répertoire donné.
Par exemple :
```
/tmp/ IN_ALL_EVENTS /home/luser/myincron.sh $@ $% $#
```
va déclencher l'execution du script `myincron.sh` sur n'importe quel évènement se produisant sur le dossier `/tmp`

Une autre solution est d'utilise les Path Units de `systemd`, mais je trouve que la configuration est un peu moins lisible pour une résultat similaire.

Une doc que je retiens parmi tout ce que j'ai lu :  https://www.linuxjournal.com/content/linux-filesystem-events-inotify

**Reply by lecorguille** [reply]
Il y a 2 stratégies: 
 - Soit on déclenche quelque chose quand un fichier arrive
 - Soit on sonde régulièrement l'arborescence à la recherche de nouveaux candidats

**Reply by FanchTheSystem** [reply]
On peux aussi mettre en place les deux strategies, en tout cas pour my on utilise les deux. (en gros avec 3 fichier systemd, un service qui lance le workflow, un timer qui lance le service toute les 10 minutes et une unit "path" qui detecte les changements)

L'interet d'avoir les 2 strategie c'est d'eviter de louper des changements. Certain evenement ne sont pas toujours recu par la systeme en fonction de la configuration du filesystem (par exemple noatime ou autre option de mount du genre)

**Reply by lecorguille** [reply]
Contrairement à My qui doit être le plus réactif possible, on peut se permettre de ne sonder que tous les 1/4 d'heure ou même plus.

A discuter

**Reply by drosofff** [reply]
Oui, en ce moment on peut même sonder tous les mois ;-)

**Comment by drosofff** [comment]
Je pense qu'un petit mail à bjoern@gruenings.eu  et Wolfgang Meier s'impose. Ils le font, ce serait bête de ne pas s'en servir.

Par ailleurs, je vais demander à ce que vous soyez invités sur le slack veg-igem.slack.com #covid9, je pense que c'est là que peuvent se débattre ce genre de questions (aussi, je ne dis pas qu'on en parle pas ailleurs).

**Reply by lecorguille** [reply]
Comme tu veux !
